internal/rbd/qos.go

```diff
@@ -0,0 +1,281 @@
+
+		return err
+	}
+
+	// if provide qosPerGB and baseVolSize, we will set qos based on capacity,
+	// otherwise, we only set base qos limit.
+	if qos.perGBLimit != "" && rbdVol.BaseVolSize != "" {
+		perGBLimit, err := strconv.ParseInt(qos.perGBLimit, 10, 64)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to parse %s: %s. %v", qos.perGBLimitType, qos.perGBLimit, err)
+
+			return err
+		}
+
+		baseVolSize, err := strconv.ParseInt(rbdVol.BaseVolSize, 10, 64)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to parse %s: %s. %v", baseVolSizeBytes, rbdVol.BaseVolSize, err)
+
+			return err
+		}
+
+		if rbdVol.VolSize <= baseVolSize {
+			rbdVol.QosParameters[qos.baseLimitType] = qos.baseLimit
+		} else {
+			capacityQos := (rbdVol.VolSize - baseVolSize) / int64(oneGB) * perGBLimit
+			finalQosLimit := baseLimit + capacityQos
+			rbdVol.QosParameters[qos.baseLimitType] = strconv.FormatInt(finalQosLimit, 10)
+		}
+	} else {
+		rbdVol.QosParameters[qos.baseLimitType] = qos.baseLimit
+	}
+
+	return nil
+}
+
+func (rv *rbdVolume) SaveQOS(
+	ctx context.Context,
+	scParams map[string]string,
+) error {
+	needSaveQosParameters := map[string]string{
+		baseReadIops:            baseQosReadIopsLimit,
+		baseWriteIops:           baseQosWriteIopsLimit,
+		baseReadBytesPerSecond:  baseQosReadBpsLimit,
+		baseWriteBytesPerSecond: baseQosWriteBpsLimit,
+		readIopsPerGB:           readIopsPerGBLimit,
+		writeIopsPerGB:          writeIopsPerGBLimit,
+		readBpsPerGB:            readBpsPerGBLimit,
+		writeBpsPerGB:           writeBpsPerGBLimit,
+		baseVolSizeBytes:        baseQosVolSize,
+	}
+	for k, v := range scParams {
+		if param, ok := needSaveQosParameters[k]; ok {
+			if v != "" {
+				err := rv.SetMetadata(param, v)
+				if err != nil {
+					log.ErrorLog(ctx, "failed to save qos. %s: %s, %v", k, v, err)
+
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func getRbdImageQOS(
+	ctx context.Context,
+	rbdVol *rbdVolume,
+) (map[string]qosSpec, error) {
+	QosParams := map[string]struct {
+		rbdQosType      string
+		rbdQosPerGBType string
+	}{
+		baseQosReadIopsLimit:  {readIopsLimit, readIopsPerGBLimit},
+		baseQosWriteIopsLimit: {writeIopsLimit, writeIopsPerGBLimit},
+		baseQosReadBpsLimit:   {readBpsLimit, readBpsPerGBLimit},
+		baseQosWriteBpsLimit:  {writeBpsLimit, writeBpsPerGBLimit},
+	}
+	rbdQosParameters := make(map[string]qosSpec)
+	for k, param := range QosParams {
+		baseLimit, err := rbdVol.GetMetadata(k)
+		if errors.Is(err, librbd.ErrNotFound) {
+			// if base qos dose not exist, skipping.
+			continue
+		} else if err != nil {
+			log.ErrorLog(ctx, "failed to get metadata: %s. %v", k, err)
+
+			return nil, err
+		}
+		perGBLimit, err := rbdVol.GetMetadata(param.rbdQosPerGBType)
+		if errors.Is(err, librbd.ErrNotFound) {
+			// rbdQosPerGBType does not exist, set it empty.
+			perGBLimit = ""
+		} else if err != nil {
+			log.ErrorLog(ctx, "failed to get metadata: %s. %v", param.rbdQosPerGBType, err)
+
+			return nil, err
```

Comment:
i think we dont need if and else if check here. we can have single check `err !=nil && !errors.Is(err, librbd.ErrNotFound)`

Reply:
updated